File: vietmap.py
import time
from selenium import webdriver
from selenium.webdriver.firefox.options import Options
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC

opts = Options()
# opts.add_argument("--headless")

driver = webdriver.Firefox(options=opts)
# Chỉnh kích thước cửa sổ 1920x1080
driver.set_window_size(1920, 1080)

driver.get("https://quanlyxe.vn/")

# Chờ tối đa 10 giây cho ô tìm kiếm xuất hiện
search_username = WebDriverWait(driver, 10).until(
    EC.presence_of_element_located((By.NAME, "UserName"))
)
print("Đã tìm kiếm xong!")

inp_pas = driver.find_element(By.NAME, "Password")
inp_login = driver.find_element(By.NAME, "login")
search_username.send_keys("hoaphat")
inp_pas.send_keys("hpl2021")
time.sleep(0.5)
inp_login.click()

current_car = '29ld31377'
vietmap = {'29ld31538' :'210002',
           '29h95648': '180285',
           '29ld31356': '212968',
           '29ld31377': '231347',
           }
#id biển xe hiện tại
current_car_id = vietmap[current_car]
# Sau khi login, chờ trang load xong lần nữa
WebDriverWait(driver, 10).until(
    lambda d: d.execute_script("return document.readyState") == "complete"
)

# Chờ đến khi window.statuses["vec_210002"] khác None data 4h 10h 48h
value = WebDriverWait(driver, 10).until(
    lambda d: d.execute_script(
    f'return window.statuses && window.statuses["vec_{current_car_id}"] !== undefined ? window.statuses["vec_{current_car_id}"] : null;')
)

print(f"Giá trị của vec_{current_car}:", value[16])

# Đợi phần tử xuất hiện và click
row = WebDriverWait(driver, 10).until(
    EC.element_to_be_clickable((By.ID, f"vec{current_car_id}"))
)
driver.execute_script("arguments[0].click();", row)

popup = WebDriverWait(driver, 10).until(
    EC.presence_of_element_located((By.CSS_SELECTOR,
        ".ol-popup.default.ol-popup-bottom.ol-popup-center.hasclosebox.shadow.visible"))
)

# Chụp ảnh phần tử popup
popup.screenshot("Bao_cao_xe.png")
print("✅ Đã lưu ảnh popup thành popup.png")
driver.quit()

File: vietmap.js
import { Builder, By, Key, until } from "selenium-webdriver";
import firefox from "selenium-webdriver/firefox.js";

async function run_vietmap(car) {
  // Cấu hình Firefox
  let options = new firefox.Options();
  options.addArguments("--headless"); // Bỏ comment nếu muốn chạy ẩn

  let driver = await new Builder()
    .forBrowser("firefox")
    .setFirefoxOptions(options)
    .build();

  try {
    // Chỉnh cửa sổ 1920x1080
    await driver.manage().window().setRect({ width: 1920, height: 1080 });

    // Mở trang web
    await driver.get("https://quanlyxe.vn/");

    // Chờ tối đa 10 giây cho ô nhập UserName
    let searchUsername = await driver.wait(
      until.elementLocated(By.name("UserName")),
      10000
    );
    console.log("Đã tìm thấy ô UserName!");

    let inpPas = await driver.findElement(By.name("Password"));
    let inpLogin = await driver.findElement(By.name("login"));

    await searchUsername.sendKeys("hoaphat");
    await inpPas.sendKeys("hpl2021");
    await driver.sleep(500); // time.sleep(0.5)
    await inpLogin.click();

    let current_car = car;
    let vietmap = {
      "29LD31538": "210002",
      "29H95648": "180285",
      "29LD31356": "212968",
      "29LD31377": "231347",
    };

    let current_car_id = vietmap[current_car];

    // Chờ đến khi xe hiện ra và click
    let row = await driver.wait(
      until.elementLocated(By.id(`vec${current_car_id}`)),
      10000
    );
    await driver.executeScript("arguments[0].click();", row);
    console.log("✅ Đã click vào xe, đang chờ popup...");

    // Chờ popup xuất hiện
    let popup = await driver.wait(
      until.elementLocated(
        By.css(
          ".ol-popup.default.ol-popup-bottom.ol-popup-center.hasclosebox.shadow.visible"
        )
      ),
      15000
    );
    console.log("✅ Popup đã xuất hiện");
    
    // Đợi thêm để popup hoàn toàn load xong và ổn định vị trí
    await driver.sleep(2000);
    
    // Đợi cho đến khi popup có nội dung và không còn thay đổi
    await driver.wait(async () => {
        try {
            const content = await popup.getText();
            return content && content.length > 0;
        } catch {
            return false;
        }
    }, 5000);
    
    console.log("✅ Popup đã load hoàn toàn");

    // Chụp ảnh popup
    await popup.takeScreenshot().then(function (image, err) {
      import("fs").then(fs => fs.writeFileSync("Bao_cao_xe.png", image, "base64"));
    });
    console.log("✅ Đã lưu ảnh popup thành Bao_cao_xe.png");
  } catch (err) {
    console.error("❌ Lỗi:", err);
  } finally {
    await driver.quit();
  }
}

export { run_vietmap };
